### PR TITLE
feat: introduce dashboard saved filter views

### PR DESCRIPTION
**Pull request by xMort**
JIRA: LX-421
risk: low

---

> [!IMPORTANT]
> Please, **don't forget to run `rush change`** for the commits that introduce **new features** 🙏

---

Refer to [documentation](https://github.com/gooddata/gooddata-ui-sdk/blob/master/dev_docs/continuous_integration.md) to see how to run checks and tests in the pull request. This is the list of the most used commands:

```
extended test - backstop
```

```
extended test - tiger-cypress - integrated
extended test - tiger-cypress - isolated
extended test - tiger-cypress - record
```

### CODE REVIEW

File: libs/sdk-ui-dashboard/src/presentation/filterBar/filterBar/FiltersConfigurationPanel.tsx [hunk 1]
```diff
@@ -1,172 +1,152 @@
...
 import { ConfigurationBubble } from "../../widget/common/configuration/ConfigurationBubble.js";
 import { metaActions } from "../../../model/store/meta/index.js";
-import { FormattedMessage } from "react-intl";
-import { useTheme } from "@gooddata/sdk-ui-theme-provider";
 
-const bubbleAlignPoints: IAlignPoint[] = [
-    { align: "tr tl", offset: { x: 0, y: 5 } },
-    { align: "tl tr", offset: { x: 0, y: 5 } },
-];
+const BUBBLE_ALIGN_POINTS: IAlignPoint[] = [{ align: "tc br", offset: { x: 4, y: -2 } }];
```
Comment (xMort) [inline libs/sdk-ui-dashboard/src/presentation/filterBar/filterBar/FiltersConfigurationPanel.tsx:19]: There was huge blog of code cut&pasted for each configuration property. I extracted it to generic component and reused it for each toggle item.

File: libs/sdk-backend-spi/src/workspace/dashboards/index.ts [hunk 2]
```diff
@@ -488,6 +490,38 @@ export interface IWorkspaceDashboardsService {
...
+    /**
+     * Set a filter view identified by the provided ref as the default one.
+     * The other filter views for the same dashboard that are marked as default ones will be unmarked.
+     *
+     * @param ref - ref of the filter view that must be set as default.
+     * @param isDefault - determine if filter view identified by the provided ref should be marked as a default
+     *      one. If yes, any existing filter view for the same user and dashboard will be marked as non default
+     *      as only one can be set as default at the same time.
+     */
+    setFilterViewAsDefault(ref: ObjRef, isDefault: boolean): Promise<void>;
```
Comment (ivan-nejezchleb) [inline libs/sdk-backend-spi/src/workspace/dashboards/index.ts:524]: if it will be called with isDefault: false it will result in NO default for dashboard?

Comment (xMort) [reply libs/sdk-backend-spi/src/workspace/dashboards/index.ts:524]: Yes. It is ok to not have any default set and just use this on fly when you want to change filters on the dashboard.

File: libs/sdk-ui-dashboard/src/model/commandHandlers/dashboard/initializeDashboardHandler/index.ts [hunk 3]
```diff
@@ -187,14 +217,17 @@ function* loadExistingDashboard(
...
         PromiseFnReturnType<typeof loadAccessibleDashboardList>,
     ] = yield all(calls);
 
     const {
-        dashboard,
+        dashboard: loadedDashboard,
         references: { insights },
     } = dashboardWithReferences;
 
+    const dashboard = applyDefaultFilterView(loadedDashboard, filterViews, config);
```
Comment (ivan-nejezchleb) [inline libs/sdk-ui-dashboard/src/model/commandHandlers/dashboard/initializeDashboardHandler/index.ts:229]: is not this code executed also during target dashboard load when drill to dashboard?

File: libs/sdk-ui-dashboard/src/model/commandHandlers/filterContext/filterViewHandler.ts [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+export function* applyFilterViewHandler(ctx: DashboardContext, cmd: ApplyFilterView): SagaIterator<void> {
+    const filterView: PromiseFnReturnType<typeof findFilterView> = yield call(
+        findFilterView,
+        cmd.payload.ref,
+    );
+
+    if (filterView) {
+        yield put(
+            changeFilterContextSelectionByParams({
```
Comment (ivan-nejezchleb) [inline libs/sdk-ui-dashboard/src/model/commandHandlers/filterContext/filterViewHandler.ts:107]: is it working with current AF config in state if it is not provided in params?

File: libs/sdk-ui-dashboard/src/model/commandHandlers/index.ts [hunk 5]
```diff
@@ -184,4 +191,9 @@ export const DefaultCommandHandlers: {
     "GDC.DASH/CMD.RENDER.ASYNC.REQUEST": requestAsyncRenderHandler,
     "GDC.DASH/CMD.RENDER.ASYNC.RESOLVE": resolveAsyncRenderHandler,
     "GDC.DASH/CMD.AUTOMATIONS.REFRESH": refreshAutomationsHandlers,
+    "GDC.DASH/CMD.FILTER_CONTEXT.FILTER_VIEW.SAVE": saveFilterViewHandler,
+    "GDC.DASH/CMD.FILTER_CONTEXT.FILTER_VIEW.DELETE": deleteFilterViewHandler,
+    "GDC.DASH/CMD.FILTER_CONTEXT.FILTER_VIEW.APPLY": applyFilterViewHandler,
+    "GDC.DASH/CMD.FILTER_CONTEXT.FILTER_VIEW.DEFAULT_STATUS_CHANGE": setFilterViewAsDefaultHandler,
+    "GDC.DASH/CMD.FILTER_CONTEXT.FILTER_VIEW.RELOADS": reloadFilterViewsHandler,
```
Comment (ivan-nejezchleb) [inline libs/sdk-ui-dashboard/src/model/commandHandlers/index.ts:198]: just .RELOAD?

File: libs/sdk-ui-dashboard/src/presentation/filterBar/filterBar/filterViews/FilterViewsList.tsx [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+                        </div>
+                    </Typography>
+                </div>
+                <div className="configuration-category">
+                    {filterViews && filterViews.length > 0 ? (
+                        filterViews.map((filterView) => (
+                            <FilterListItem
+                                key={objRefToString(filterView.ref)}
+                                item={filterView}
+                                onApply={() => {
```
Comment (ivan-nejezchleb) [inline libs/sdk-ui-dashboard/src/presentation/filterBar/filterBar/filterViews/FilterViewsList.tsx:146]: should not be these handlers defined via useCallbacks?

**Review by ivan-nejezchleb — CHANGES_REQUESTED** [review]

File: libs/sdk-ui-dashboard/src/model/commandHandlers/filterContext/filterViewHandler.ts (hunk 4 above)
Comment (xMort) [reply libs/sdk-ui-dashboard/src/model/commandHandlers/filterContext/filterViewHandler.ts:107]: Seems like it is probably not using it. I plan to look into this. I would leave it as is in this pull request and will fix it in the next one.

File: libs/sdk-ui-dashboard/src/model/commandHandlers/filterContext/filterViewHandler.ts [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+        yield put(reloadFilterViews());
+        yield put(filterViewDefaultStatusChangeSucceeded(ctx, updatedFilterView, cmd.correlationId));
+    } catch (error) {
+        defaultErrorHandler(error);
+        yield put(filterViewDefaultStatusChangeFailed(ctx, updatedFilterView));
+    }
+}
+
+export function* reloadFilterViewsHandler(ctx: DashboardContext): SagaIterator<void> {
+    const filterViews: PromiseFnReturnType<typeof loadFilterViews> = yield call(loadFilterViews, ctx);
```
[the remark this answers is not in the document]
Comment (xMort) [reply libs/sdk-ui-dashboard/src/model/commandHandlers/filterContext/filterViewHandler.ts:142]: I wrapped this handler in try catch as the others.

File: libs/sdk-ui-dashboard/src/model/commandHandlers/index.ts [hunk 8]
```diff
@@ -184,4 +191,9 @@ export const DefaultCommandHandlers: {
     "GDC.DASH/CMD.RENDER.ASYNC.REQUEST": requestAsyncRenderHandler,
     "GDC.DASH/CMD.RENDER.ASYNC.RESOLVE": resolveAsyncRenderHandler,
     "GDC.DASH/CMD.AUTOMATIONS.REFRESH": refreshAutomationsHandlers,
+    "GDC.DASH/CMD.FILTER_CONTEXT.FILTER_VIEW.SAVE": saveFilterViewHandler,
+    "GDC.DASH/CMD.FILTER_CONTEXT.FILTER_VIEW.DELETE": deleteFilterViewHandler,
+    "GDC.DASH/CMD.FILTER_CONTEXT.FILTER_VIEW.APPLY": applyFilterViewHandler,
+    "GDC.DASH/CMD.FILTER_CONTEXT.FILTER_VIEW.DEFAULT_STATUS_CHANGE": setFilterViewAsDefaultHandler,
```
[the remark this answers is not in the document]
Comment (xMort) [reply libs/sdk-ui-dashboard/src/model/commandHandlers/index.ts:197]: Good point. Changed. I refactor these command keys few times already.

File: libs/sdk-ui-dashboard/src/presentation/filterBar/filterBar/filterViews/FilterViewsList.tsx (hunk 6 above)
Comment (xMort) [reply libs/sdk-ui-dashboard/src/presentation/filterBar/filterBar/filterViews/FilterViewsList.tsx:146]: I don't think it is necessary. This way everything would have to be wrapped. Unless it is used some heavy hook or performance is bad, it does not need to be wrapped imho. Also, this is called in `.map` so I would have to move the whole logic down to the component, provide `onClose` to it and leave the component to handle the creation of the callbacks. It look nicer to have it here on the higher layer.

File: libs/sdk-ui-dashboard/src/model/commandHandlers/filterContext/filterViewHandler.ts (hunk 4 above)
Comment (xMort) [reply libs/sdk-ui-dashboard/src/model/commandHandlers/filterContext/filterViewHandler.ts]: I think I saw the usage of `changeFilterContextSelectionByParams` in reset filter feature originally. I looked into it some more and changed the command to `changeFilterContextSelection` and I'm not messing with configs any more.

**Review by ivan-nejezchleb — APPROVED** [review]

**Pull request merged**